apps/sms/test/marionette/lib/storage.js

```diff
@@ -57,15 +27,29 @@
          * pre-populate storage with.
          */
         setContactsStorage: function(contacts) {
-          client.executeScript(function(contacts) {
-            var STORAGE_NAME = 'contactsDB';
-
-            window.wrappedJSObject.TestStorages.setStorage(STORAGE_NAME, {
-              contacts: contacts
-            });
+          this.setStorage('contactsDB', {
+            contacts: contacts || []
+          });
+        },
 
-            window.wrappedJSObject.TestStorages.setStorageReady(STORAGE_NAME);
-          }, [contacts || []]);
+        /**
+         * Sets pre-populated custom storage.
+         * @param {string} storageName Name of the custom storage.
+         * @param {*} storage Custom storage data.
+         * @private
+         */
+        setStorage: function(storageName, storage) {
+          client.scope({ context: 'chrome' }).executeScript(
+            function(storage, storageName) {
+              Services.obs.addObserver(function(window) {
+                window.wrappedJSObject.TestStorages.set(
+                  storageName,
+                  Components.utils.cloneInto(storage, window.wrappedJSObject)
+                );
+              }, 'test-storage-ready', false);
```

Comment:
>  if we don't unregister we keep the storage in memory.

Well, imo, it's not a big deal for the integration tests - I believe memory should be freed once suite is finished. And as I explain below this event is called for several different frames, so we can safely unsubscribe only when once all frames fired this event and we set storage for all of them.

> if we call setStorage once again after a relaunch, we will have the handler called twice.

Yes and this should not break test as storage will be overridden by the last set storage - we can consider this as  temporary way to override storage on the next startup :) We don't have such cases, but as always - once one need this feature - he'll make it better.

However I agree that the way our injected scripts work currently is confusing and not clear. We just follow the way marionette content script plugin suggests us - register observer for document-element-inserted event. That means we call it for keyboard, homescreen, Messages main frame and shim host - not a big deal and somewhat correct that all apps has the same API output during one test case, but I see that this is called twice for every document element - that what I'd like to investigate and filed bug for that - https://bugzilla.mozilla.org/show_bug.cgi?id=1223350